### ISSUE TITLE
bootMer + parallel not finding functions or objects

### ISSUE THREAD

**Issue by KaiserDominici** [lme4/lme4#435]
Reproducible example, from the help page:

`fm01ML <- lmer(Yield ~ 1|Batch, Dyestuff, REML = FALSE)`
```
mySumm <- function(.) {
s <- sigma(.)
c( beta = getME(., "beta"), sigma = s, sig01 = unname(s * getME(., "theta")))
}
```
I load the `parallel` library and save a log file when making a new cluster:
```
library(parallel)
cl <- makeCluster(rep("localhost", 4), outfile = 'log.txt')
showConnections(all = T)
  description                            class      mode  text     isopen   can read can write
0 "stdin"                                "terminal" "r"   "text"   "opened" "yes"    "no"     
1 "stdout"                               "terminal" "w"   "text"   "opened" "no"     "yes"    
2 "stderr"                               "terminal" "w"   "text"   "opened" "no"     "yes"    
3 "<-<address>:<port>"                   "sockconn" "a+b" "binary" "opened" "yes"    "yes"    
4 "<-<address>:<port>"                   "sockconn" "a+b" "binary" "opened" "yes"    "yes"    
5 "<-<address>:<port>"                   "sockconn" "a+b" "binary" "opened" "yes"    "yes"    
6 "<-<address>:<port>"                   "sockconn" "a+b" "binary" "opened" "yes"    "yes"
```
(I have censored my internal address, but as you can see four connections have been opened). Now:
```
boo01 <- bootMer(fm01ML, mySumm, nsim = 100, verbose = T, parallel = "snow", ncpus = 4, cl = cl) 
Warning message:
In bootMer(fm01ML, mySumm, nsim = 100, verbose = T, parallel = "snow",  :
  some bootstrap runs failed (100/100)
```
If I open the log, this is what I see:
```
starting worker pid=8116 on localhost:<port> at 10:01:33.729
starting worker pid=4088 on localhost:<port> at 10:01:34.061
starting worker pid=7708 on localhost:<port> at 10:01:34.415
starting worker pid=7284 on localhost:<port> at 10:01:34.759
    1 :List of 2
 $ message: chr "could not find function \"getME\""
 $ call   : language FUN(refit(x, ss[[i]]))
 - attr(*, "class")= chr [1:3] "simpleError" "error" "condition"
    2 :List of 2
```
And it goes on like that for each simulation.

I first encountered this problem with a custom `mySumm` function that used an object in my work environment as `newdata` and the log said that the object was not found.
```
sessionInfo()
R version 3.3.2 (2016-10-31)
Platform: i386-w64-mingw32/i386 (32-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

locale:
[1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252   
[3] LC_MONETARY=English_United Kingdom.1252 LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.1252    

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] bindrcpp_0.2    merTools_0.3.0  arm_1.9-3       MASS_7.3-45     lme4_1.1-13     Matrix_1.2-7.1 
 [7] knitr_1.17      lubridate_1.6.0 dplyr_0.7.2     purrr_0.2.2.2   readr_1.1.1     tidyr_0.6.3    
[13] tibble_1.3.3    ggplot2_2.2.1   tidyverse_1.1.1 magrittr_1.5   

loaded via a namespace (and not attached):
 [1] jsonlite_1.5     reshape2_1.4.2   nloptr_1.0.4     gtable_0.2.0     minqa_1.2.4     
 [6] hms_0.3          xml2_1.1.1       htmltools_0.3.6  forcats_0.2.0    stringr_1.2.0   
[11] haven_1.1.0      broom_0.4.2      splines_3.3.2    cellranger_1.1.0 lattice_0.20-34 
[16] htmlwidgets_0.9  plyr_1.8.4       blme_1.0-4       mime_0.5         pkgconfig_2.0.1 
[21] R6_2.2.2         shiny_1.0.3      digest_0.6.12    colorspace_1.3-2 DT_0.2          
[26] stringi_1.1.5    lazyeval_0.2.0   labeling_0.3     abind_1.4-5      httr_1.2.1      
[31] bindr_0.1        xtable_1.8-2     munsell_0.4.3    psych_1.7.5      modelr_0.1.1    
[36] Rcpp_0.12.12     readxl_1.0.0     highr_0.6        coda_0.19-1      assertthat_0.2.0
[41] tools_3.3.2      foreign_0.8-67   httpuv_1.3.5     mvtnorm_1.0-6    scales_0.4.1    
[46] glue_1.1.1       rlang_0.1.1      mnormt_1.5-5     nlme_3.1-128     rvest_0.3.2     
[51] grid_3.3.2`
```

Thank you for any help,

k.

**Comment by bbolker** [comment]
Reproducible.  FWIW `parallel="multicore"` works fine (of course it does, that's how we've tested ...)

**Comment by bbolker** [comment]
Running `clusterEvalQ(cl,library("lme4"))` appears to fix the problem for me. Can you confirm that it works for you?  If so, I'll add it to the documentation.

**Comment by KaiserDominici** [comment]
Hi Ben,

Thank you for looking into this. I can confirm that `clusterEvalQ(cl,library("lme4"))` fixes this for me. The process exits without error.
```
attr(boo01, which = "bootFail")
[1] 0
```

Same goes for my custom `mySumm` function, I exported the object to the cluster and it worked.

Best wishes,

k.

**Comment by bbolker** [comment]
Does this note in the documentation seem sufficiently clear/complete?

 >  If you are using \code{parallel="snow"}, you will need to run
\code{clusterEvalQ(cl,library("lme4"))} to make sure that the
\code{lme4} package is loaded on all of the workers; you may
additionally need to use \code{\link[parallel]{clusterExport}}
if you are using a summary function that calls any objects
from the environment.

**Comment by KaiserDominici** [comment]
I would perhaps say:

> If you are using \code{parallel="snow"}, you will need to run
\code{clusterEvalQ(cl,library("lme4"))} before calling \code{bootMer}
to make sure ...

So that it is clear that you should export/evaluate to the cluster _before_ running `bootMer`, or is it so obvious it is not worth stating explicitly?

**Comment by bbolker** [comment]
OK.  Fixed in 16da968794 .
